### DISCUSSION TITLE
Is http basic authentication supported for "git-tags" datasource on dependency lookup?

### DISCUSSION THREAD

**Discussion by herndlm** [renovatebot/renovate#10504] [Request Help] [answered]
Hi,

we have a composer.json that uses VCS repositories. Initially Renovate was not able to create any PRs because (http basic) authentication was not configured for those. After configurating it via hostrule for the datasource "packagist" Renovate is able to create PRs for other dependencies (previously it would fail creating the lockfile). But apparently the VCS deps cannot be resolved although I tried to configure authentication for them but was not successful.

The error
```
DEBUG: Datasource unknown error (repository=<REDACTED_REPOSITORY>)
       "datasource": "git-tags",
       "lookupName": "http://<REDACTED>/<REDACTED>.git",
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             "http://<REDACTED>/<REDACTED>.git"
           ],
           "format": "utf-8"
         },
         "message": "fatal: could not read Username for 'http://<REDACTED>': No such device or address\n",
         "stack": "Error: fatal: could not read Username for 'http://<REDACTED>': No such device or address\n\n    at Object.action (/usr/src/app/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:38:28)\n    at PluginStore.exec (/usr/src/app/node_modules/simple-git/src/lib/plugins/plugin-store.ts:24:29)\n    at /usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:114:40\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:111:14)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:88:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
       }
DEBUG: Failed to look up dependency <REDACTED_DEP> (repository=<REDACTED_REPOSITORY>, packageFile=composer.json, dependency=<REDACTED_DEP>)
``` 

Composer auth was configured successfully via env vars `PACKAGIST_<DOMAIN>_USERNAME` and `PACKAGIST_<DOMAIN>_PASSWORD`. I tried to do the same with `GIT__TAGS_<DOMAIN>_USERNAME` and I tried to add the domain without a hostType to my config directly.

But the debug output never shows any basic auth headers/commands although the merged config looked just fine.

Could it be that Renovate does not use the hostRules when looking up dependencies via git? I wanted to ask before trying to create a reproducable project and open a bug.

**Comment by rarkins** [comment, marked as answer]
I'm not 100% sure after a quick look. The `git-tags` datasource invokes the `git-refs` datasource, which has this call: https://github.com/renovatebot/renovate/blob/9dd254ab845d3c3be98caef4f30597b72f03d135/lib/datasource/git-refs/index.ts#L32

In turn it calls this function: https://github.com/renovatebot/renovate/blob/9dd254ab845d3c3be98caef4f30597b72f03d135/lib/util/git/url.ts#L11-L21

It looks like that token only cares about `token` but could be extended to check for `username`/`password` and insert them in the URL instead.

**Reply by herndlm** [reply]
Great, thx! I started implementing this and there might be a few more questions or things to discuss but this is part of the PR then I guess

**Reply by tbugfinder** [reply]
Hi @herndlm ,
I'm also running into the describe issue while git-tags datasource is used within a bitbucket-server environment. What's the state of the PR?
Unfortunately I wasn't able to do any token hack for the given setup.

**Reply by herndlm** [reply]
Oh, this was my first renovate PR :) and it was working for us ever since. https://github.com/renovatebot/renovate/pull/10513

There is just some other composer VCS repo config adaption needed, I'll mention it here and adapt the docs maybe later.

**Reply by viceice** [reply]
@tbugfinder please open a new discussion and describe as best as you can what are you trying and how do you configure renovate for that.

**Reply by herndlm** [reply]
Just for the record - the additional adaption I mentioned is now documented at https://docs.renovatebot.com/modules/manager/composer/#additional-information
